allennlp/models/semantic_parsing/quarel/quarel_semantic_parser.py

```diff
@@ -0,0 +1,890 @@
+        question : Dict[str, torch.LongTensor]
+           The output of ``TextField.as_array()`` applied on the question ``TextField``. This will
+           be passed through a ``TextFieldEmbedder`` and then through an encoder.
+        table : ``Dict[str, torch.LongTensor]``
+            The output of ``KnowledgeGraphField.as_array()`` applied on the table
+            ``KnowledgeGraphField``.  This output is similar to a ``TextField`` output, where each
+            entity in the table is treated as a "token", and we will use a ``TextFieldEmbedder`` to
+            get embeddings for each entity.
+        world : ``List[QuarelWorld]``
+            We use a ``MetadataField`` to get the ``World`` for each input instance.  Because of
+            how ``MetadataField`` works, this gets passed to us as a ``List[QuarelWorld]``,
+        actions : ``List[List[ProductionRuleArray]]``
+            A list of all possible actions for each ``World`` in the batch, indexed into a
+            ``ProductionRuleArray`` using a ``ProductionRuleField``.  We will embed all of these
+            and use the embeddings to determine which action to take at each timestep in the
+            decoder.
+        target_action_sequences : torch.Tensor, optional (default=None)
+           A list of possibly valid action sequences, where each action is an index into the list
+           of possible actions.  This tensor has shape ``(batch_size, num_action_sequences,
+           sequence_length)``.
+        """
+
+        table_text = table['text']
+
+        self._debug_count -= 1
+
+        # (batch_size, question_length, embedding_dim)
+        embedded_question = self._question_embedder(question)
+        question_mask = util.get_text_field_mask(question).float()
+        num_question_tokens = embedded_question.size(1)
+
+        # (batch_size, num_entities, num_entity_tokens, embedding_dim)
+        embedded_table = self._question_embedder(table_text, num_wrapping_dims=1)
+        table_mask = util.get_text_field_mask(table_text, num_wrapping_dims=1).float()
+
+        batch_size, num_entities, num_entity_tokens, _ = embedded_table.size()
+
+        # entity_types: one-hot tensor with shape (batch_size, num_entities, num_types)
+        # entity_type_dict: Dict[int, int], mapping flattened_entity_index -> type_index
+        # These encode the same information, but for efficiency reasons later it's nice
+        # to have one version as a tensor and one that's accessible on the cpu.
+        entity_types, entity_type_dict = self._get_type_vector(world, num_entities, embedded_table)
+
+        if self._use_entities:
+            if self._entity_encoder is not None:
+                # (batch_size, num_entities, embedding_dim)
+                encoded_table = self._entity_encoder(embedded_table, table_mask)
+                # (batch_size, num_entities, num_neighbors)
+                neighbor_indices = self._get_neighbor_indices(world, num_entities, encoded_table)
```

Comment:
Are you actually using neighbors at all in this model?  What for?

Reply:
No, the neighbors are always empty, so this will not do anything interesting. In fact I don't use any entity_encoder in the current models, so these more mostly artifacts left over from starting with the WikiTables code.

Reply:
I just started removing some of this old stuff, but I don't have a test to run to make sure I'm not breaking anything, and I can't easily add one because I don't have any test fixtures to use.

Reply:
I added some very basic tests now, but they're so far only verifying that the model trains and saves.